src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,205 @@
+                    "Please wait until the upload finishes.")));
+        return;
+    }
+
+    isUploading = true;
+    channel->addMessage(makeSystemMessage(QString("Started upload...")));
+
+    if (source->hasFormat("image/png"))
+    {
+        uploadImageToNuuls(source->data("image/png"), channel, outputTextEdit);
+    }
+    else if (source->hasFormat("text/uri-list"))
+    {
+        QStringList potientialPathsToSend =
+            QString(source->data("text/uri-list").toStdString().c_str())
+                .split("\r\n");
+
+        for (QString path : potientialPathsToSend)
+        {
+            if (path.isEmpty())
+            {
+                break;
+            }
+            else
+            {
+                if (getImageFileFormat(path) != QString())
+                {
+                    channel->addMessage(makeSystemMessage(
+                        QString("Uploading image: %1").arg(path)));
+                    QImage img = QImage(QUrl(path).toLocalFile());
+                    if (img.isNull())
+                    {
+                        channel->addMessage(makeSystemMessage(
+                            QString("Couldn't load image :(")));
+                        return;
+                    }
+                    QByteArray imageData;
+                    QBuffer buf(&imageData);
+                    buf.open(QIODevice::WriteOnly);
+                    img.save(&buf, "png");
+
+                    TypedBytes data = {imageData, "png"};
+                    uploadQueue.push(data);
+                }
+                else if (path.endsWith(".gif"))
+                {
+                    channel->addMessage(makeSystemMessage(
+                        QString("Uploading GIF: %1").arg(path)));
+                    QFile file(QUrl(path).toLocalFile());
+                    bool isOkay = file.open(QIODevice::ReadOnly);
+                    if (!isOkay)
+                    {
+                        channel->addMessage(makeSystemMessage(
+                            QString("Failed to open file. :(")));
+                        return;
+                    }
+                    TypedBytes data = {file.readAll(), "gif"};
+                    uploadQueue.push(data);
+                    file.close();
+                    // file.readAll() => might be a bit big but it /should/ work
+                }
+                else
+                {
+                    channel->addMessage(makeSystemMessage(
+                        QString("Cannot upload file: %1, not an image")
+                            .arg(path)));
+                }
+            }
+        }
+        if (uploadQueue.size())
+        {
+            uploadImageToNuuls(uploadQueue.front(), channel, outputTextEdit);
+            uploadQueue.pop();
+        }
+    }
+    else if (source->hasFormat("image/gif"))
+    {
+        TypedBytes data = {source->data("image/gif"), "gif"};
+        uploadImageToNuuls(data, channel, outputTextEdit);
+    }
+    else
```

Comment:
Jpeg images should just be uploaded and not be converted to png. Other format like tiff/whatever are fine to be converted IMO.